nilearn/decomposition/dict_learning.py

```diff
@@ -0,0 +1,256 @@
+        S = np.sqrt(np.sum(self.dict_init_ ** 2, axis=0))
+        self.dict_init_ /= S[np.newaxis, :]
+
+    def fit(self, imgs, y=None, confounds=None):
+        """Compute the mask and the ICA maps across subjects
+
+        Parameters
+        ----------
+        imgs: list of Niimg-like objects
+            See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+            Data on which PCA must be calculated. If this is a list,
+            the affine is considered the same for all.
+
+        confounds: CSV file path or 2D matrix
+            This parameter is passed to nilearn.signal.clean. Please see the
+            related documentation for details
+        """
+        # Base logic for decomposition estimators
+        DecompositionEstimator.fit(self, imgs)
+
+        if self.verbose:
+            print('[DictLearning] Loading data')
+        with mask_and_reduce(self.masker_, imgs, confounds,
+                             reduction_ratio=self.reduction_ratio,
+                             n_components=self.n_components,
+                             random_state=self.random_state,
+                             memory_level=self.memory_level,
+                             memory=self.memory,
+                             max_nbytes=self.max_nbytes) as data:
+            if self.verbose:
+                print('[DictLearning] Initializating dictionary')
+            self._init_dict(data)
+
+            if self.n_epochs < 0:
+                self.n_epochs = 1
+            # Performing more than 10 epochs would probably useless
+            if self.n_epochs > 10:
+                self.n_epochs = 10
+            n_iter = int(ceil((data.shape[1] / 10 * self.n_epochs)))
+
+            if self.verbose:
+                print('[DictLearning] Learning dictionary')
+            dictionary = self._cache(dict_learning_online,
+                                     func_memory_level=2)(
+                data.T,
+                self.n_components,
+                alpha=self.alpha,
+                n_iter=n_iter,
+                batch_size=10,
+                method='cd',
+                return_code=False,
+                dict_init=self.dict_init_,
+                verbose=max(0, self.verbose - 1),
+                random_state=self.random_state,
+                shuffle=False,
+                n_jobs=1)
+            self.components_ = self._cache(sparse_encode,
+                                           func_memory_level=2,
+                                           ignore=['n_jobs'])\
+                (data.T, dictionary, algorithm='lasso_cd', alpha=self.alpha,
+                 n_jobs=self.n_jobs).T
+        self.components_ = as_ndarray(self.components_)
+
+        # flip signs in each composant positive part is l1 larger
+        #  than negative part
+        for component in self.components_:
+            if np.sum(component[component > 0]) <\
+                    - np.sum(component[component <= 0]):
```

Comment:
If the component is sparse (lots of 0s in the background), you will always flip it.